x/metadata/keeper/migrations_v4.go

```diff
@@ -0,0 +1,169 @@
+package keeper
+
+import (
+	"fmt"
+
+	"cosmossdk.io/log"
+	storetypes "cosmossdk.io/store/types"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/types/address"
+	"github.com/cosmos/gogoproto/proto"
+
+	markertypes "github.com/provenance-io/provenance/x/marker/types"
+	"github.com/provenance-io/provenance/x/metadata/types"
+)
+
+// Migrate3To4 will update the metadata store from version 3 to version 4. This should be part of the viridian upgrade.
+func (m Migrator) Migrate3To4(ctx sdk.Context) error {
+	logger := m.keeper.Logger(ctx)
+	logger.Info("Starting migration of x/metadata from 3 to 4.")
+	if err := migrateValueOwners(ctx, newKeeper3To4(m.keeper)); err != nil {
+		logger.Error("Error migrating scope value owners.", "error", err)
+		return err
+	}
+	logger.Info("Done migrating x/metadata from 3 to 4.")
+	return nil
+}
+
+// keeper3To4I is an interface with keeper-related stuff needed to migrate metadata from 3 to 4.
+type keeper3To4I interface {
+	Logger(ctx sdk.Context) log.Logger
+	GetStore(ctx sdk.Context) storetypes.KVStore
+	Unmarshal(bz []byte, ptr proto.Message) error
+	SetScopeValueOwner(ctx sdk.Context, scopeID types.MetadataAddress, newValueOwner string) error
+}
+
+// keeper3To4 is a wrapper on the metadata Keeper with a few extra things exposed to satisfy the keeper3To4I interface.
+type keeper3To4 struct {
+	Keeper
+}
+
+// newKeeper3To4 wraps the provided metadata Keeper as a keeper3To4.
+func newKeeper3To4(kpr Keeper) keeper3To4 {
+	return keeper3To4{Keeper: kpr}
+}
+
+// GetStore returns a store for the metadata stuff.
+func (k keeper3To4) GetStore(ctx sdk.Context) storetypes.KVStore {
+	return ctx.KVStore(k.Keeper.storeKey)
+}
+
+// Unmarshal uses the Keeper's codec to unmarshal something.
+func (k keeper3To4) Unmarshal(bz []byte, ptr proto.Message) error {
+	return k.Keeper.cdc.Unmarshal(bz, ptr)
+}
+
+// V3WriteNewScope writes a new scope to state in the way v3 of the metadata module did.
+// Deprecated: Only exists to facilitate testing of the migration of the metadata module from v3 to v4.
+func (k Keeper) V3WriteNewScope(ctx sdk.Context, scope types.Scope) error {
+	store := ctx.KVStore(k.storeKey)
+	if store.Has(scope.ScopeId) {
+		return fmt.Errorf("scope %s already exists", scope.ScopeId)
+	}
+	bz, err := k.cdc.Marshal(&scope)
+	if err != nil {
+		return fmt.Errorf("could not marshal scope %s: %w", scope.ScopeId, err)
+	}
+	store.Set(scope.ScopeId, bz)
+	k.indexScope(store, &scope, nil)
+	// indexScope no longer does anything with the value owner address, but
+	// we used to have a couple index entries for it.
+	if len(scope.ValueOwnerAddress) > 0 {
+		vo := sdk.MustAccAddressFromBech32(scope.ValueOwnerAddress)
+		k1 := types.GetAddressScopeCacheKey(vo, scope.ScopeId)
+		k2 := GetValueOwnerScopeCacheKey(vo, scope.ScopeId)
+		store.Set(k1, []byte{0x01})
+		store.Set(k2, []byte{0x01})
+	}
+	return nil
+}
```

Comment:
I must exist until after the migration has been executed. It can't live in a utility package because it needs stuff private to the keeper. And it can't exist in a `_test.go` file because it's needed outside this package.

But it DOES live in a file that will be entirely deleted at some time after the migration has been executed as part of the `viridian` cleanup. So there's already a plan to delete it.